### PR TITLE
FileJournal: New journal padding mode

### PR DESCRIPTION
**Pull request by XinzeChi**
Current, because the requirement of DirectIo, the padding mode
is adding padding data for every entry.

In fact, we fetch a couple of entries as a system write as possible
as can. So we only make sure system-write content met the requirement
of DirectIO. We adding a padding entry at the end of system-write
content(for most case, this include a couple of entries).

User could set the osd_journal_padding_ahead to decide which mode.
osd_journal_padding_ahead = true is good choice for pcie-ssd.

This work is base on https://github.com/ceph/ceph/pull/5877.
Signed-off-by: Jianpeng Ma <jianpeng.ma@intel.com>
Signed-off-by: Xinze Chi <xinze@xsky.com>

### CODE REVIEW

**Comment by XinzeChi** [comment]
Benchmark result: 
4k randwrite, the iops could increase from 10600 -> 11000.

Intel(R) Xeon(R) CPU E5-2620 v2 @ 2.10GHz
1 SATA SSD (2 OSD )
32G Mem.

**Comment by XinzeChi** [comment]
@liewegas , Please review it. Thanks.

File: src/os/FileJournal.cc [hunk 1]
```diff
@@ -643,7 +644,11 @@ int FileJournal::_fdump(Formatter &f, bool simple)
       &bl,
       &seq,
       &ss);
-    if (result != SUCCESS) {
+    if (result == SKIP) {
+      seq = pre_seq;
```
Comment (liewegas) [inline src/os/FileJournal.cc]: would it be simpler to make do_read_entry not clobber &seq when returning SKIP?

File: src/os/FileJournal.cc [hunk 2]
```diff
@@ -823,10 +828,10 @@ void FileJournal::write_header_sync()
   dout(20) << __func__ << " finish" << dendl;
 }
 
-int FileJournal::check_for_full(uint64_t seq, off64_t pos, off64_t size)
+int FileJournal::check_for_full(off64_t pos, off64_t size, bool skip_full)
 {
   // already full?
-  if (full_state != FULL_NOTFULL)
+  if (!skip_full && full_state != FULL_NOTFULL)
```
Comment (liewegas) [inline src/os/FileJournal.cc]: why?

File: src/os/FileJournal.cc [hunk 3]
```diff
@@ -866,6 +871,45 @@ int FileJournal::check_for_full(uint64_t seq, off64_t pos, off64_t size)
   return -ENOSPC;
 }
 
+int FileJournal::prepare_padding_entry(bufferlist& bl, off64_t& queue_pos)
```
Comment (liewegas) [inline src/os/FileJournal.cc]: If I'm reading this right, if you have some number of entries concatenated together that aren't already aligned, you "wrap" them in a dummy header + some padding to align them?

File: src/os/FileJournal.cc (hunk 2 above)
Comment (XinzeChi) [inline src/os/FileJournal.cc]: image


we have reserve enough space for padding entry.

File: src/os/FileJournal.cc (hunk 3 above)
Comment (liewegas) [inline src/os/FileJournal.cc]: It seems like a much simpler approach would be:

change prepare_multi_write to pass the start alignment and bool is_last to prepare_single_write.  Then make prepare_single_write only pad if (1) the write is big and we care about alignment of the data payload (say, > 16KB), or (2) we are the last entry in the batch, in which case we need to add tail padding to align ourselves.

What do you think?

Comment (XinzeChi) [inline src/os/FileJournal.cc]: if they are not aligned, we would call align_bl the align them. Funcation align_bl would dp memcpy and  release some ptr. so this is not friendly to pcie ssd.
So if we use the pcie ssd, we should not use this patch.  user could set osd_journal_padding_ahead to true in pcie ssd.

**Comment by dachary** [comment]
is this still current ? If yes, it needs a rebase ;-)

**Comment by XinzeChi** [comment]
We want to focus efforts on optimizations with BlueStore. Close it.

**Pull request closed without merging**